src/main/kotlin/me/partlysanestudios/partlysaneskies/features/chat/ChatColors.kt

```diff
@@ -0,0 +1,93 @@
+//
+// Written by Su386.
+// See LICENSE for copyright and license notices.
+//
+
+package me.partlysanestudios.partlysaneskies.features.chat
+
+import me.partlysanestudios.partlysaneskies.PartlySaneSkies
+import me.partlysanestudios.partlysaneskies.utils.StringUtils.matchGroup
+import me.partlysanestudios.partlysaneskies.utils.StringUtils.removeColorCodes
+import net.minecraft.util.ChatComponentText
+import net.minecraft.util.IChatComponent
+
+object ChatColors {
+
+    fun detectColorMessage(message: IChatComponent): IChatComponent {
+        val formattedMessage = message.formattedText
+        val prefix = getPrefix(formattedMessage)
+        if (prefix.isEmpty()) return message
+
+        val color = getChatColor(prefix)
+        return if (color.isEmpty()) message
+        else ChatComponentText(insertColor(formattedMessage, color))
+    }
+
+    fun detectNonMessage(message: IChatComponent): IChatComponent {
+        if (!PartlySaneSkies.config.colorNonMessages) return message
+
+        val formattedMessage = message.formattedText
+        if (!formattedMessage.contains("§r§7: ")) return message
+
+        if (formattedMessage.startsWith("§dTo ") || formattedMessage.startsWith("§dFrom ")) {
+            return message
+        }
+
+        val unformattedMessage = message.unformattedText
+        val containsRankNames = PartlySaneSkies.RANK_NAMES.any { unformattedMessage.contains(it) }
+
+        return if (containsRankNames) message
+        else ChatComponentText(insertColor(formattedMessage, "§r"))
+    }
+
+    fun getChatColor(prefix: String): String = when (prefix.lowercase()) {
+        "party" -> {
+            if (!PartlySaneSkies.config.colorPartyChat) ""
+            else if (PartlySaneSkies.config.visibleColors) "§6"
+            else "§9"
+        }
+        "guild" -> {
+            if (!PartlySaneSkies.config.colorGuildChat) ""
+            else if (PartlySaneSkies.config.visibleColors) "§a"
+            else "§2"
+        }
+        "officer" -> {
+            if (!PartlySaneSkies.config.colorOfficerChat) ""
+            else "§3"
+        }
+        "to", "from" -> {
+            if (!PartlySaneSkies.config.colorPrivateMessages) ""
+            else "§d"
+        }
+        "co-op" -> {
+            if (!PartlySaneSkies.config.colorCoopChat) ""
+            else "§b"
+        }
+        else -> ""
+    }
+
+    private val prefixPattern = "(?<chat>Party|Guild|Officer|To|From|Co-op) >?".toPattern()
```

Comment:
This match fails bc it checks the whole chat message, so you would need a .* at the end